internal/cimfs/cim_reader_windows.go

```diff
@@ -0,0 +1,1070 @@
+}
+
+// A File is a file or directory within an open CIM.
+type File struct {
+	r    *CimFsReader
+	name string
+	sr   streamReader
+	ino  *inode
+}
+
+type inode struct {
+	id          format.FileID
+	file        format.File
+	linkTable   []byte
+	streamTable []byte
+}
+
+type streamReader struct {
+	stream     format.Stream
+	off        uint64
+	pe         format.PeImage
+	pemappings []format.PeImageMapping
+	peinit     bool
+}
+
+// A Stream is an alternate data stream of a file within a CIM.
+type Stream struct {
+	c     *CimFsReader
+	r     streamReader
+	fname string
+	name  string
+}
+
+func readBin(r io.Reader, v interface{}) error {
+	err := binary.Read(r, binary.LittleEndian, v)
+	if err == io.EOF {
+		err = io.ErrUnexpectedEOF
+	}
+	return err
+}
+
+func validateHeader(h *format.CommonHeader) error {
+	if !bytes.Equal(h.Magic[:], format.MagicValue[:]) {
+		return errors.New("not a cim file")
+	}
+	if h.Version.Major != format.CurrentVersion.Major {
+		return fmt.Errorf("unsupported cim version. cim version %v expected version %v", h.Version, format.CurrentVersion)
+	}
+	return nil
+}
+
+func loadRegionSet(rs *format.RegionSet, imagePath string, reg []region) (int, error) {
+	for i := 0; i < int(rs.Count); i++ {
+		name := fmt.Sprintf("region_%v_%d", rs.ID, i)
+		rf, err := os.Open(filepath.Join(imagePath, name))
+		if err != nil {
+			return 0, err
+		}
+		reg[i].f = rf
+		fi, err := rf.Stat()
+		if err != nil {
+			return 0, err
+		}
+		reg[i].size = uint64(fi.Size())
+		var rh format.RegionHeader
+		err = readBin(rf, &rh)
+		if err != nil {
+			return 0, fmt.Errorf("reading region header %s: %s", name, err)
+		}
+		err = validateHeader(&rh.Common)
+		if err != nil {
+			return 0, fmt.Errorf("validating region header %s: %s", name, err)
+		}
+	}
+	return int(rs.Count), nil
+}
+
+func readFilesystemHeader(f *os.File) (format.FilesystemHeader, error) {
+	var fsh format.FilesystemHeader
+	err := readBin(f, &fsh)
+	if err != nil {
+		return fsh, fmt.Errorf("reading filesystem header: %s", err)
+	}
+	err = validateHeader(&fsh.Common)
+	if err != nil {
+		return fsh, fmt.Errorf("validating filesystem header: %s", err)
+	}
+	return fsh, nil
+}
+
+// Open opens a CIM file for read access.
+func Open(p string) (_ *CimFsReader, err error) {
+	defer func() {
+		if err != nil {
+			err = &OpError{Cim: p, Op: "open", Err: err}
+		}
+	}()
+	f, err := os.Open(p)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	fsh, err := readFilesystemHeader(f)
+	if err != nil {
+		return nil, err
+	}
+	parents := make([]format.RegionSet, fsh.ParentCount)
+	err = readBin(f, parents)
+	if err != nil {
+		return nil, fmt.Errorf("reading parent region sets: %s", err)
+	}
+	regionCount := int(fsh.Regions.Count)
+	for i := range parents {
+		regionCount += int(parents[i].Count)
+	}
+	if regionCount == 0 || regionCount > 0x10000 {
+		return nil, fmt.Errorf("invalid region count %d", regionCount)
+	}
+	c := &CimFsReader{
```

Comment:
Any value in making a `NewCimFsReader` func?

Reply:
As of now this is the only place where we create the `CimFsReader` instance and I don't expect any external code to create the CimFsReader by creating this struct. So I will keep this as it is for now but we can make a new func later if required.

Reply:
sounds good!